File: apac/20160828/B.py
# coding=utf8

import sys
import numpy as np

def fast_pow(x, y):
    r = 1
    base = x
    while y != 0:
        if y & 1 != 0:
            r = r * base % 1000000007L
        base = base * base % 1000000007L
        y /= 2
    return r

def solve(kase):
    a, b, n, k = scanf("%d %d %d %d")
    cnt = 0
    for i in xrange(1, n+1):
        for j in xrange(1, n+1):
            if i != j:
                if (fast_pow(i, a) + fast_pow(j, b)) % k == 0:
                    cnt += 1
                    cnt = cnt % 1000000007L
    print "Case #%d: %d" % (kase, cnt % 1000000007L)

if __name__ == '__main__':
    if len(sys.argv) >= 2:
        sys.stdin = open(sys.argv[1], 'r')
    if len(sys.argv) >= 3:
        sys.stdout = open(sys.argv[2], 'w')
    
    from scanf import *

    kase, = scanf("%d\n")
    for i in xrange(1, kase+1):
        solve(i)

File: apac/20160828/B.cpp
#include <stdio.h>
#include <string.h>
#include <string>
#include <unordered_map>
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

#define MOD 1000000007L
#define LL long long

LL fast_pow(LL x, LL y, LL k) {
	LL r = 1; x %= k;
	while(y) {
		if(y & 1) r = r * x % k;
		x = x * x % k;
		y >>= 1;
	}
	return r % k;
}

void solve(int kase){
	LL res = 0, a, b, n, k;
	scanf("%lld %lld %lld %lld", &a, &b, &n, &k);
	for(LL i=1; i<=n; i+=1) {
		for(LL j=1; j<=n; j+=1) {
			if(i != j) {
				if((fast_pow(i, a, k) + fast_pow(j, b, k)) % k == 0) {
					res = (res + 1) % MOD;
				}
			}
		}
	}
	printf("Case #%d: %lld\n", kase, res);
}

int main(int argc, char *argv[]){
	if(argc >= 2) {
		freopen(argv[1], "r", stdin);
	}else{
		freopen("B.in", "r", stdin);	
	}
	int kase;
	scanf("%d\n", &kase);
	for(int i=1; i<=kase; i+=1){
		solve(i);	
	}
	return 0;
}

